.ci/magician/cmd/generate_downstream.go

```diff
@@ -0,0 +1,338 @@
+	},
+}
+
+func listGDEnvironmentVariables() string {
+	var result string
+	for i, ev := range gdEnvironmentVariables {
+		result += fmt.Sprintf("\t%2d. %s\n", i+1, ev)
+	}
+	return result
+}
+
+func execGenerateDownstream(baseBranch, command, repo, version, ref string, gh GithubClient, rnr *exec.Runner, ctlr *source.Controller) {
+	if baseBranch == "" {
+		baseBranch = "main"
+	}
+	upstreamRepo, scratchRepo, err := cloneRepo(baseBranch, repo, version, ctlr)
+	if err != nil {
+		fmt.Println("Error cloning repo: ", err)
+		os.Exit(1)
+	}
+
+	mmLocalPath := filepath.Join(rnr.GetCWD(), "..", "..")
+	mmCopyPath := filepath.Join(mmLocalPath, "..", fmt.Sprintf("mm-%s-%s-%s", repo, version, command))
+	if _, err := rnr.Run("cp", []string{"-rp", mmLocalPath, mmCopyPath}, nil); err != nil {
+		fmt.Println("Error copying magic modules: ", err)
+		os.Exit(1)
+	}
+
+	if err := rnr.PushDir(mmCopyPath); err != nil {
+		fmt.Println("Error changing directory to copied magic modules: ", err)
+		os.Exit(1)
+	}
+
+	if err := setConfig(rnr); err != nil {
+		fmt.Println("Error setting config: ", err)
+		os.Exit(1)
+	}
+
+	mmRepo := &source.Repo{
+		Name: "magic-modules",
+		Path: mmCopyPath,
+	}
+	commitMessage, err := branchAndCommitMessage(mmRepo, scratchRepo, command, ref, rnr, ctlr)
+	if err != nil {
+		fmt.Println("Error getting branch and commit message: ", err)
+		os.Exit(1)
+	}
+
+	if err := runMake(upstreamRepo, command, rnr); err != nil {
+		fmt.Println("Error running make: ", err)
+		os.Exit(1)
+	}
+
+	if err := pushCommit(upstreamRepo, scratchRepo, baseBranch, command, commitMessage, ref, gh, rnr, ctlr); err != nil {
+		fmt.Println("Error pushing commit: ", err)
+		os.Exit(1)
+	}
+}
+
+func cloneRepo(baseBranch, repo, version string, ctlr *source.Controller) (*source.Repo, *source.Repo, error) {
+	upstreamRepo := &source.Repo{
+		Title:  repo,
+		Branch: baseBranch,
+	}
+	switch repo {
+	case "terraform":
+		if version == "ga" {
+			upstreamRepo.Name = "terraform-provider-google"
+			upstreamRepo.Version = provider.GA
+		} else if version == "beta" {
+			upstreamRepo.Name = "terraform-provider-google-beta"
+			upstreamRepo.Version = provider.Beta
+		} else {
+			return nil, nil, fmt.Errorf("unrecognized version %s", version)
+		}
+		upstreamRepo.Owner = "hashicorp"
+	case "terraform-google-conversion":
+		upstreamRepo.Name = "terraform-google-conversion"
+		upstreamRepo.Owner = "GoogleCloudPlatform"
+	case "tf-oics":
+		if upstreamRepo.Branch == "main" {
+			upstreamRepo.Branch = "master"
+		}
+		upstreamRepo.Name = "docs-examples"
+		upstreamRepo.Owner = "terraform-google-modules"
+	case "tf-cloud-docs":
+		fmt.Println(repo, " is no longer available.")
+		return nil, nil, nil
+	default:
+		return nil, nil, fmt.Errorf("unrecognized repo %s", repo)
+	}
+	ctlr.SetPath(upstreamRepo)
+	if err := ctlr.Clone(upstreamRepo); err != nil {
+		return nil, nil, err
+	}
+	scratchRepo := &source.Repo{
```

Comment:
nit: its kind of hard to follow source repo and scratch repo's state since they are interconnected. Notably the branch changes on the checked out branch which then makes the other repos 'state' invalid. It's albiet a bit confusing from my perspective.

Reply:
I'm not sure what you mean here. The scratch repo refers to the fork owned by the magician, while the downstream repo is the actual repo.

Reply:
Yes, but source controller (in my mental model) is used to clone and track a repository. It's a bit weird how we clone it and the metadata of the object says the state is on a certain branch when they are both tracking the same physical repo location.